### ISSUE TITLE
torch.compile() drops the performance of validation

### ISSUE THREAD

**Issue by jiahong-fu** [pytorch/pytorch#100061]
### 🐛 Describe the bug

I encountered that after adding torch.compile(model), the performance of the model on the verification set has been declining as the training progresses.
![image](https://user-images.githubusercontent.com/55650657/234467056-5434eaf2-1c7c-43ac-bbab-adeaa26a4660.png)

When I remove torch.compile(model), the model is back to normal:
epoch 1/1000, train: loss=0.0767, val: psnr=27.5813, 1.4m 1.4m/24.2h
epoch 2/1000, train: loss=0.0438, val: psnr=28.2431, 1.6m 3.0m/25.3h
epoch 3/1000, train: loss=0.0428, val: psnr=28.2909, 1.6m 4.6m/25.5h
epoch 4/1000, train: loss=0.0417, val: psnr=28.4744, 1.6m 6.2m/25.9h
epoch 5/1000, train: loss=0.0406, val: psnr=28.7785, 1.6m 7.8m/26.1h
epoch 6/1000, train: loss=0.0404, val: psnr=28.5917, 1.6m 9.5m/26.3h
epoch 7/1000, train: loss=0.0395, val: psnr=28.6923, 1.6m 11.1m/26.3h
epoch 8/1000, train: loss=0.0393, val: psnr=28.9986, 1.5m 12.6m/26.2h
epoch 9/1000, train: loss=0.0382, val: psnr=28.8534, 1.7m 14.2m/26.4h

### Error logs

Those two things do not work. You can reproduce the similar result with this [repro](https://github.com/XieQi2015/F-Conv/tree/main/MinistExp).

You can implement with the simple case, by adding the model = torch.compile(model) below [here](https://github.com/XieQi2015/F-Conv/blob/c55c4184a0cf6e7a30be62217c781264853a0515/MinistExp/Rotated_MNIST_simpleCase_Main.py#L46).

After training 30 epochs, the model begin to test. If you add the torch.compile, the test accuracy will not increase. While for the version without torch.compile is normal.

### Minified repro

env TORCHDYNAMO_REPRO_AFTER="aot" TORCHDYNAMO_REPRO_LEVEL=4

System prompts the body is too long, so I unload the repro.py in  [here](https://drive.google.com/file/d/1cD1NV3ucDpHxn9unu04kmlTnjNVAgWRt/view)

### Versions

Collecting environment information...
PyTorch version: 2.0.0
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.1 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.31

Python version: 3.10.11 (main, Apr 20 2023, 19:02:41) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.4.0-42-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 11.1.74
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 4090
GPU 1: NVIDIA GeForce RTX 4090
GPU 2: NVIDIA GeForce RTX 4090
GPU 3: NVIDIA GeForce RTX 4090
GPU 4: NVIDIA GeForce RTX 4090
GPU 5: NVIDIA GeForce RTX 4090
GPU 6: NVIDIA GeForce RTX 4090
GPU 7: NVIDIA GeForce RTX 4090

Nvidia driver version: 525.105.17
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Byte Order:                      Little Endian
Address sizes:                   48 bits physical, 48 bits virtual
CPU(s):                          256
On-line CPU(s) list:             0-254
Off-line CPU(s) list:            255
Thread(s) per core:              1
Core(s) per socket:              64
Socket(s):                       2
NUMA node(s):                    8
Vendor ID:                       AuthenticAMD
CPU family:                      25
Model:                           1
Model name:                      AMD EPYC 7763 64-Core Processor
Stepping:                        1
Frequency boost:                 enabled
CPU MHz:                         1502.643
CPU max MHz:                     2450.0000
CPU min MHz:                     1500.0000
BogoMIPS:                        4900.14
Virtualization:                  AMD-V
L1d cache:                       2 MiB
L1i cache:                       2 MiB
L2 cache:                        32 MiB
L3 cache:                        256 MiB
NUMA node0 CPU(s):               0-15,128-143
NUMA node1 CPU(s):               16-31,144-159
NUMA node2 CPU(s):               32-47,160-175
NUMA node3 CPU(s):               48-63,176-191
NUMA node4 CPU(s):               64-79,192-207
NUMA node5 CPU(s):               80-95,208-223
NUMA node6 CPU(s):               96-111,224-239
NUMA node7 CPU(s):               112-127,240-254
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Full AMD retpoline, IBPB conditional, IBRS_FW, STIBP always-on, RSB filling
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr wbnoinvd arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload vgif umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca

Versions of relevant libraries:
[pip3] natten==0.14.6+torch200cu118
[pip3] numpy==1.23.5
[pip3] pytorch-ssim==0.1
[pip3] torch==2.0.0
[pip3] torchaudio==2.0.0
[pip3] torchvision==0.15.0
[pip3] triton==2.0.0
[conda] blas                      1.0                         mkl  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.4.0           h06a4308_640  
[conda] mkl-service               2.4.0           py310h7f8727e_0  
[conda] mkl_fft                   1.3.1           py310hd6ae3a3_0  
[conda] mkl_random                1.2.2           py310h00e6091_0  
[conda] natten                    0.14.6+torch200cu118          pypi_0    pypi
[conda] numpy                     1.23.5          py310hd5efca6_0  
[conda] numpy-base                1.23.5          py310h8e6c178_0  
[conda] pytorch                   2.0.0           py3.10_cuda11.8_cudnn8.7.0_0    pytorch
[conda] pytorch-cuda              11.8                 h7e8668a_3    pytorch-nightly
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] pytorch-ssim              0.1                      pypi_0    pypi
[conda] torchaudio                2.0.0               py310_cu118    pytorch
[conda] torchtriton               2.0.0                     py310    pytorch
[conda] torchvision               0.15.0              py310_cu118    pytorch

**Comment by ezyang** [comment]
try https://github.com/pytorch/pytorch/pull/100226

**Comment by anijain2305** [comment]
The issue is because of incorrect decomp for `elu_backward`. If I comment out this line https://github.com/pytorch/pytorch/blob/main/torch/_decomp/decompositions.py#L118, the accuracy is good again.

Unfortunately, I could not use minifiers here, and had to do a lot of manual debugging to reach this stage.

**Comment by anijain2305** [comment]
Verified that https://github.com/pytorch/pytorch/pull/100284 fixes this problem

**Comment by jiahong-fu** [comment]
Sry, the problem still exists. As @anijain2305 , if i comment out https://github.com/pytorch/pytorch/blob/main/torch/_decomp/decompositions.py#L118, the training accuracy is good again. But for the test (the model begin to test after 30 epochs) accuracy keeps getting worse. @ezyang 

epoch 0 | train accuracy: 71.16666666666667
epoch 1 | train accuracy: 92.28333333333333
epoch 2 | train accuracy: 94.27499999999999
epoch 3 | train accuracy: 95.85000000000001
epoch 4 | train accuracy: 95.93333333333334
epoch 5 | train accuracy: 96.41666666666666
epoch 6 | train accuracy: 96.09166666666667
epoch 7 | train accuracy: 97.08333333333333
epoch 8 | train accuracy: 96.85833333333333
epoch 9 | train accuracy: 97.15833333333333
epoch 10 | train accuracy: 97.26666666666667
epoch 11 | train accuracy: 97.15
epoch 12 | train accuracy: 97.68333333333334
epoch 13 | train accuracy: 97.30833333333334
epoch 14 | train accuracy: 97.48333333333333
epoch 15 | train accuracy: 97.53333333333333
epoch 16 | train accuracy: 98.05
epoch 17 | train accuracy: 97.85000000000001
epoch 18 | train accuracy: 97.66666666666667
epoch 19 | train accuracy: 97.78333333333333
epoch 20 | train accuracy: 97.85833333333333
epoch 21 | train accuracy: 97.98333333333333
epoch 22 | train accuracy: 98.08333333333333
epoch 23 | train accuracy: 97.65833333333333
epoch 24 | train accuracy: 97.74166666666667
epoch 25 | train accuracy: 98.10833333333333
epoch 26 | train accuracy: 97.99166666666666
epoch 27 | train accuracy: 98.04166666666667
epoch 28 | train accuracy: 97.91666666666666
epoch 29 | train accuracy: 98.3
epoch 30 | train accuracy: 99.075
epoch 31 | train accuracy: 99.55833333333334
epoch 32 | train accuracy: 99.59166666666667
The test accuracy: 98.544
The test error is 1.45600
The best error is 1.36600
=================================================
epoch 33 | train accuracy: 99.61666666666666
The test accuracy: 98.424
The test error is 1.57600
The best error is 1.36600
=================================================
epoch 34 | train accuracy: 99.71666666666667
The test accuracy: 98.19
The test error is 1.81000
The best error is 1.36600
=================================================
epoch 35 | train accuracy: 99.65833333333333
The test accuracy: 97.758
The test error is 2.24200
The best error is 1.36600
=================================================
epoch 36 | train accuracy: 99.78333333333333
The test accuracy: 97.178
The test error is 2.82200
The best error is 1.36600
=================================================
epoch 37 | train accuracy: 99.64166666666667
The test accuracy: 96.41
The test error is 3.59000
The best error is 1.36600
=================================================
epoch 38 | train accuracy: 99.59166666666667
The test accuracy: 94.772
The test error is 5.22800
The best error is 1.36600
=================================================
epoch 39 | train accuracy: 99.65
The test accuracy: 94.242
The test error is 5.75800
The best error is 1.36600
=================================================
epoch 40 | train accuracy: 99.69166666666666
The test accuracy: 92.704
The test error is 7.29600
The best error is 1.36600
=================================================
epoch 41 | train accuracy: 99.76666666666667
The test accuracy: 88.388
The test error is 11.61200
The best error is 1.36600
=================================================
epoch 42 | train accuracy: 99.73333333333333
The test accuracy: 86.69
The test error is 13.31000

**Comment by ezyang** [comment]
Just to confirm, is this most recent result with https://github.com/XieQi2015/F-Conv/tree/main/MinistExp or https://github.com/XieQi2015/F-Conv/blob/c55c4184a0cf6e7a30be62217c781264853a0515/MinistExp/Rotated_MNIST_simpleCase_Main.py#L46 ?

**Comment by jiahong-fu** [comment]
The result of https://github.com/XieQi2015/F-Conv/tree/main/MinistExp. You can reproduce with this case, by adding the model = torch.compile(model) below [here](https://github.com/XieQi2015/F-Conv/blob/c55c4184a0cf6e7a30be62217c781264853a0515/MinistExp/Rotated_MNIST_simpleCase_Main.py#L46).

**Comment by jiahong-fu** [comment]
Is there any new progress on this issue?

**Comment by ezyang** [comment]
I hadn't gotten a chance to run your example. I think the first thing to check is an ablation: does the test accuracy improve with backend="eager", backend="aot_eager" and backend="aot_eager_decomp_partition".

**Comment by ezyang** [comment]
OK it ran and I was able to reproduce the problem, I'll run the ablations overnight

**Comment by jiahong-fu** [comment]
What can I do to fix this?

**Comment by anijain2305** [comment]
Yes the guards are coming from the newly setup `filter` and `bias` in F_Conv `train` method.

When we guard on attributes of `nn.Modules` (here filter and bias are buffers), we recompile for each `eval` (and this is correct because each `eval` changes the filter and bias in this model). The following snippet shows the different guards for first and second test run. 

![image](https://user-images.githubusercontent.com/13822661/236580452-ef1fe083-e463-4740-9554-f107e726e709.png)

Sam's work should cover this.

**Comment by CalinLucian** [comment]
Hello everybody I can confirm using torch_compile=True in Trainer() in HuggingFace does indeed speed up 40% the training time but the eval_f1 score drops by 0.10/0.15. With torch_compile=False I get 0.10/0.15 bigger f1_score. The dataset is the same, I checked that multiple times.

**Comment by BenjaminBossan** [comment]
Hi, I came here via #105863 and would like to know what the status is.

This is a big problem for Hugging Face PEFT, as it seems to make the use of any `if` statement inside of `forward` impossible. The issue persists in PyTorch v2.1.0. A minimal reproducer:

```python
import torch
from torch import nn

class Model(nn.Module):
    def __init__(self):
        super().__init__()
        self.switch = True

    def forward(self, X):
        if self.switch:
            return torch.zeros_like(X)
        return torch.ones_like(X)

def test_it(do_compile=False):
    X = torch.arange(5).float()
    model = Model()

    if do_compile:
        model = torch.compile(model)
        print("compiled the model")

    outputs0 = model(X)
    model.switch = False
    outputs1 = model(X)

    torch.testing.assert_close(outputs0, torch.zeros(5))
    torch.testing.assert_close(outputs1, torch.ones(5))

if __name__ == '__main__':
    print(torch.__version__)
    test_it(do_compile=False)  # passes
    print("tests without compile passed")
    test_it(do_compile=True)  # fails
    print("tests with compilie passed")
```

**Comment by wbigat** [comment]
It seems that this problem still exists on the latest main version.  This is my case,  #111785 .  

I wonder what the plan is and will this be fixed in the future?  @anijain2305 @ezyang

**Comment by tugsbayasgalan** [comment]
Is this still an issue @ezyang ?

**Comment by ezyang** [comment]
yes

**Comment by anijain2305** [comment]
This is fixed on main by `guard_nn_modules=True`. Closing the issue for now.
